mlrun/datastore/google_cloud_storage.py

```diff
@@ -103,8 +128,41 @@ def put(self, key, data, append=False):
             f.write(data)
 
     def upload(self, key, src_path):
-        path = self._make_path(key)
-        self.filesystem.put_file(src_path, path, overwrite=True)
+        file_size = os.path.getsize(src_path)
+        united_path = self._make_path(key)
+
+        # Multiple upload limitation recommendations as described in
+        # https://cloud.google.com/storage/docs/multipart-uploads#storage-upload-object-chunks-python
+
+        if file_size < self.chunk_size:
+            self.filesystem.put_file(src_path, united_path, overwrite=True)
+            return
+
+        bucket = self.storage_client.bucket(self.endpoint)
+        bucket_retention_policy = bucket._properties.get("retentionPolicy")
+        if bucket_retention_policy or bucket.object_retention_mode:
+            self.filesystem.put_file(src_path, united_path, overwrite=True)
+            return
+        blob = bucket.blob(key.strip("/"))
+
+        if (
+            blob.retention_expiration_time
+            or blob.temporary_hold
+            or blob.event_based_hold
```

Comment:
From the client library code, it doesn't seem like these can be defined immediately after creating the blob, but maybe I'm missing something.